File: web/src/types/agent.ts
export type LLMConfig = {
  id: number
  name: string
  provider: string
  base_url?: string
  model: string
}

export type Tool = {
  name: string
  description: string
}

export type EventRecord = {
  id: string
  name: string
  status: 'running' | 'done' | 'error'
  input?: string
  output?: string
  agent?: string
  agent_path?: string
}

export type SubagentEventRecord = {
  id: string
  name?: string | null
  path: string
  status: 'running' | 'done' | 'error' | 'completed' | 'failed' | 'cancelled' | string
  input?: string | null
  error?: string | null
}

export type AgentSession = {
  id: number
  title?: string | null
  config_id?: number | null
  system_prompt?: string | null
  temperature?: number | null
  max_tokens?: number | null
  enable_1m_context?: boolean | null
  enable_rag?: boolean | null
  tools?: string[] | null
  mcp_servers?: string[] | null
  agent_configs?: Record<string, number | null> | null
  created_at: string
  updated_at: string
}

export type AgentSessionTemplate = {
  id: number
  name: string
  config_id?: number | null
  system_prompt?: string | null
  temperature?: number | null
  max_tokens?: number | null
  enable_1m_context?: boolean | null
  enable_rag?: boolean | null
  tools?: string[] | null
  mcp_servers?: string[] | null
  agent_configs?: Record<string, number | null> | null
  created_at: string
  updated_at: string
}

export type MCPServer = {
  id: number
  name: string
  type: 'command' | 'sse' | 'streamable-http'
  command?: string
  args?: string[]
  env?: Record<string, string>
  url?: string
  enabled: boolean
  created_at: string
  updated_at: string
}

export type AgentMessage = {
  id: number
  session_id: number
  role: 'user' | 'assistant' | 'tool'
  content: string
  name?: string
  tool_call_id?: string
  tool_calls?: any[]
  extra_metadata?: {
    rag_sources?: string[]
    tool_events?: EventRecord[]
    mcp_events?: EventRecord[]
    subagent_events?: SubagentEventRecord[]
  }
  input_tokens?: number
  cached_input_tokens?: number
  output_tokens?: number
  model?: string
  created_at: string
}

export type SessionFile = {
  path: string
  size: number
  modified_at: string
  type: 'file' | 'directory'
}

File: server/schemas/agent.py
"""Agent session schemas."""

from datetime import datetime
from typing import Dict, List, Optional

from pydantic import BaseModel, ConfigDict


class AgentRunRequest(BaseModel):
    """Agent run request schema."""

    config_id: int
    input: str
    session_id: Optional[int] = None
    system_prompt: Optional[str] = None
    temperature: Optional[float] = None
    max_tokens: Optional[int] = None
    enable_1m_context: Optional[bool] = None
    enable_rag: Optional[bool] = False
    tools: Optional[List[str]] = None
    mcp_servers: Optional[List[str]] = None
    agent_configs: Optional[Dict[str, Optional[int]]] = None
    agent_mcp_servers: Optional[Dict[str, Optional[List[str]]]] = None


class ToolResponse(BaseModel):
    """Tool information response schema."""

    name: str
    description: str


class SubAgentResponse(BaseModel):
    """Subagent information response schema."""

    name: str
    description: str


class AgentSessionCreate(BaseModel):
    """Agent session creation schema."""

    title: Optional[str] = None


class AgentSessionUpdate(BaseModel):
    """Agent session update schema."""

    title: Optional[str] = None
    config_id: Optional[int] = None
    system_prompt: Optional[str] = None
    temperature: Optional[float] = None
    max_tokens: Optional[int] = None
    enable_1m_context: Optional[bool] = None
    enable_rag: Optional[bool] = None
    tools: Optional[List[str]] = None
    mcp_servers: Optional[List[str]] = None
    agent_configs: Optional[Dict[str, Optional[int]]] = None
    agent_mcp_servers: Optional[Dict[str, Optional[List[str]]]] = None


class AgentSessionResponse(BaseModel):
    """Agent session response schema."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    title: Optional[str] = None
    config_id: Optional[int] = None
    system_prompt: Optional[str] = None
    temperature: Optional[float] = None
    max_tokens: Optional[int] = None
    enable_1m_context: Optional[bool] = None
    enable_rag: Optional[bool] = False
    tools: Optional[List[str]] = None
    mcp_servers: Optional[List[str]] = None
    agent_configs: Optional[Dict[str, Optional[int]]] = None
    agent_mcp_servers: Optional[Dict[str, Optional[List[str]]]] = None
    created_at: datetime
    updated_at: datetime

class AgentSessionTemplateCreate(BaseModel):
    """Agent session settings template creation schema."""

    name: str
    config_id: Optional[int] = None
    system_prompt: Optional[str] = None
    temperature: Optional[float] = None
    max_tokens: Optional[int] = None
    enable_1m_context: Optional[bool] = None
    enable_rag: Optional[bool] = None
    tools: Optional[List[str]] = None
    mcp_servers: Optional[List[str]] = None
    agent_configs: Optional[Dict[str, Optional[int]]] = None
    agent_mcp_servers: Optional[Dict[str, Optional[List[str]]]] = None


class AgentSessionTemplateUpdate(BaseModel):
    """Agent session settings template update schema."""

    name: Optional[str] = None
    config_id: Optional[int] = None
    system_prompt: Optional[str] = None
    temperature: Optional[float] = None
    max_tokens: Optional[int] = None
    enable_1m_context: Optional[bool] = None
    enable_rag: Optional[bool] = None
    tools: Optional[List[str]] = None
    mcp_servers: Optional[List[str]] = None
    agent_configs: Optional[Dict[str, Optional[int]]] = None
    agent_mcp_servers: Optional[Dict[str, Optional[List[str]]]] = None


class AgentSessionTemplateResponse(BaseModel):
    """Agent session settings template response schema."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    config_id: Optional[int] = None
    system_prompt: Optional[str] = None
    temperature: Optional[float] = None
    max_tokens: Optional[int] = None
    enable_1m_context: Optional[bool] = None
    enable_rag: Optional[bool] = False
    tools: Optional[List[str]] = None
    mcp_servers: Optional[List[str]] = None
    agent_configs: Optional[Dict[str, Optional[int]]] = None
    agent_mcp_servers: Optional[Dict[str, Optional[List[str]]]] = None
    created_at: datetime
    updated_at: datetime

class AgentMessageResponse(BaseModel):
    """Agent message response schema."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    session_id: int
    role: str
    content: str
    name: Optional[str] = None
    tool_call_id: Optional[str] = None
    tool_calls: Optional[List[dict]] = None
    extra_metadata: Optional[dict] = None
    input_tokens: Optional[int] = None
    cached_input_tokens: Optional[int] = None
    output_tokens: Optional[int] = None
    model: Optional[str] = None
    created_at: datetime

class SessionFileResponse(BaseModel):
    """Session file response schema."""

    path: str
    size: int
    modified_at: datetime
    type: str  # "file" or "directory"
